nixops/resources/vault_approle.py

```diff
@@ -0,0 +1,218 @@
+import nixops.util
+import nixops.resources
+import nixops.vault_common
+import json
+import requests
+
+
+class VaultApproleDefinition(nixops.resources.ResourceDefinition):
+    """Definition of a Vault Approle"""
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    @classmethod
+    def get_resource_type(cls):
+        return "vaultApprole"
+
+    def show_type(self):
+        return "{0}".format(self.get_type())
+
+
+class VaultApproleState(nixops.resources.DiffEngineResourceState):
+    """State of a Vault Approle"""
+
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    vault_token = nixops.util.attr_property("vaultToken", None)
+    _reserved_keys = [ "vaultAddress", "roleName", "roleId", "bindSecretId"
+                        "secretIdBoundCidrs", "tokenBoundCidrs", "policies"
+                        "secretIdNumUses", "secretIdTtl", "tokenNumUses"
+                        "tokenMaxTtl", "tokenTtl", "period", "enableLocalSecretIds"
+                        "tokenType", "secretId", "cidrList"
+                      ]
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    def __init__(self, depl, name, id):
+        nixops.resources.DiffEngineResourceState.__init__(self, depl, name, id)
+        self.role_name = self._state.get('roleName', None)
+        self.role_id = self._state.get('roleId', None)
+        self.vault_address = self._state.get('vaultAddress', None)
+        self.secret_id = self._state.get('secretId', None)
+
+    def show_type(self):
+        s = super(VaultApproleState, self).show_type()
+        return s
+
+    @property
+    def resource_id(self):
+        path = '/v1/auth/approle/role/'
+        return self.vault_address + path + self.role_name if self.role_name else None
+
+    def get_definition_prefix(self):
+        return "resources.vaultApprole."
+
+    def get_physical_spec(self):
+        path = '/v1/auth/approle/role/'
+        return {'url': self.vault_address + path + self.role_name } if self.role_name else {}
+
+    def prefix_definition(self, attr):
+        return {('resources', 'vaultApprole'): attr}
+
+    def _connect(self, vault_token):
```

Comment:
It doesn't seem that _connect does any actual connection, we typically use a connect method to create an http client and cache it for later uses. And btw any reason for not using an actual Vault python library such as https://github.com/hvac/hvac

Reply:
Connect is basically to test if the token is actually valid and can create role it is not lile other _conect methods in ec2 pr datadog.
we can use hvoc but it is an overkill. Also, I was not able to package it in nixpkgs since it relies on another package where the maintainer is freezing a given other package (lot of packages :p) to a version and fail for to build with the version in nixpkgs.